File: packages/contracts/websocket.py
"""WebSocket Protocol Inbound and Outbound Frame Contracts."""

from pydantic import BaseModel, ConfigDict, Field

from .enums import AudioStreamType, MeetingStatus, WSClientMessageType, WSServerMessageType
from .rest import ParticipantContract


class BaseWSFrame(BaseModel):
    model_config = ConfigDict(
        populate_by_name=True,
        validate_assignment=True,
        extra="forbid",
    )


# --- Client to Server Frames ---


class WSClientJoinFrame(BaseWSFrame):
    type: WSClientMessageType = WSClientMessageType.JOIN
    ticket: str = Field(..., description="Single-use WebSocket session ticket issued by API")
    participant_id: str


class WSClientSetLanguageFrame(BaseWSFrame):
    type: WSClientMessageType = WSClientMessageType.SET_LISTENING_LANGUAGE
    listening_language: str = Field(
        ..., min_length=3, max_length=3, description="ISO-639-3 target listening language"
    )


class WSClientChatMessageFrame(BaseWSFrame):
    type: WSClientMessageType = WSClientMessageType.CHAT_MESSAGE
    text: str = Field(..., min_length=1, max_length=2000)


class WSClientQueryAssistantFrame(BaseWSFrame):
    type: WSClientMessageType = WSClientMessageType.QUERY_ASSISTANT
    query_id: str
    question: str = Field(..., min_length=1, max_length=1000)


class WSClientPingFrame(BaseWSFrame):
    type: WSClientMessageType = WSClientMessageType.PING
    timestamp_ms: int


# --- Server to Client Frames ---


class WSServerRoomStateFrame(BaseWSFrame):
    type: WSServerMessageType = WSServerMessageType.ROOM_STATE
    state_version: int
    status: MeetingStatus
    participants: list[ParticipantContract]


class WSServerParticipantJoinedFrame(BaseWSFrame):
    type: WSServerMessageType = WSServerMessageType.PARTICIPANT_JOINED
    state_version: int
    participant: ParticipantContract


class WSServerParticipantLeftFrame(BaseWSFrame):
    type: WSServerMessageType = WSServerMessageType.PARTICIPANT_LEFT
    state_version: int
    participant_id: str


class WSServerCaptionFrame(BaseWSFrame):
    type: WSServerMessageType = WSServerMessageType.CAPTION_UPDATE
    source_segment_id: str = Field(..., description="Immutable lineage source segment ID")
    speaker_id: str
    source_language: str
    target_language: str
    text: str
    is_final: bool
    start_ms: int
    end_ms: int


class WSServerAudioTrackFrame(BaseWSFrame):
    type: WSServerMessageType = WSServerMessageType.AUDIO_TRACK_PUBLISHED
    track_sid: str
    language: str
    stream_type: AudioStreamType


class WSServerAssistantFrame(BaseWSFrame):
    type: WSServerMessageType = WSServerMessageType.ASSISTANT_RESPONSE
    query_id: str
    answer: str
    citations: list[str] = []
    action_items: list[str] = []


class WSServerPongFrame(BaseWSFrame):
    type: WSServerMessageType = WSServerMessageType.PONG
    timestamp_ms: int


class WSServerErrorFrame(BaseWSFrame):
    type: WSServerMessageType = WSServerMessageType.ERROR
    code: str
    message: str

File: packages/contracts/src/websocket.ts
import { AudioStreamType, MeetingStatus, WSClientMessageType, WSServerMessageType } from "./enums";
import { ParticipantContract } from "./rest";

// --- Client to Server Frames ---

export interface WSClientJoinFrame {
  type: WSClientMessageType.JOIN;
  ticket: string;
  participant_id: string;
}

export interface WSClientSetLanguageFrame {
  type: WSClientMessageType.SET_LISTENING_LANGUAGE;
  listening_language: string;
}

export interface WSClientChatMessageFrame {
  type: WSClientMessageType.CHAT_MESSAGE;
  text: string;
}

export interface WSClientQueryAssistantFrame {
  type: WSClientMessageType.QUERY_ASSISTANT;
  query_id: string;
  question: string;
}

export interface WSClientPingFrame {
  type: WSClientMessageType.PING;
  timestamp_ms: number;
}

export type WSClientFrame =
  | WSClientJoinFrame
  | WSClientSetLanguageFrame
  | WSClientChatMessageFrame
  | WSClientQueryAssistantFrame
  | WSClientPingFrame;

// --- Server to Client Frames ---

export interface WSServerRoomStateFrame {
  type: WSServerMessageType.ROOM_STATE;
  state_version: number;
  status: MeetingStatus;
  participants: ParticipantContract[];
}

export interface WSServerParticipantJoinedFrame {
  type: WSServerMessageType.PARTICIPANT_JOINED;
  state_version: number;
  participant: ParticipantContract;
}

export interface WSServerParticipantLeftFrame {
  type: WSServerMessageType.PARTICIPANT_LEFT;
  state_version: number;
  participant_id: string;
}

export interface WSServerCaptionFrame {
  type: WSServerMessageType.CAPTION_UPDATE;
  source_segment_id: string;
  speaker_id: string;
  source_language: string;
  target_language: string;
  text: string;
  is_final: boolean;
  start_ms: number;
  end_ms: number;
}

export interface WSServerAudioTrackFrame {
  type: WSServerMessageType.AUDIO_TRACK_PUBLISHED;
  track_sid: string;
  language: string;
  stream_type: AudioStreamType;
}

export interface WSServerAssistantFrame {
  type: WSServerMessageType.ASSISTANT_RESPONSE;
  query_id: string;
  answer: string;
  citations: string[];
  action_items: string[];
}

export interface WSServerPongFrame {
  type: WSServerMessageType.PONG;
  timestamp_ms: number;
}

export interface WSServerErrorFrame {
  type: WSServerMessageType.ERROR;
  code: string;
  message: string;
}

export type WSServerFrame =
  | WSServerRoomStateFrame
  | WSServerParticipantJoinedFrame
  | WSServerParticipantLeftFrame
  | WSServerCaptionFrame
  | WSServerAudioTrackFrame
  | WSServerAssistantFrame
  | WSServerPongFrame
  | WSServerErrorFrame;
